### PR TITLE
[SMTChecker] Support user types

### LINKED ISSUE #11966: [SMTChecker] Support user defined types

### PR DESCRIPTION
**Pull request by leonardoalt**
Fix https://github.com/ethereum/solidity/issues/11966

Still need to add more tests, but code can be reviewed

### CODE REVIEW

File: test/libsolidity/smtCheckerTests/userTypes/user_abi_2.sol [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+contract C {
+	function f(bytes memory data) public pure {
+		T x = abi.decode(data, (T));
+		T y = abi.decode(data, (T));
+		assert(T.unwrap(x) == T.unwrap(y)); // should hold
+		assert(T.unwrap(x) != T.unwrap(y)); // should fail
+	}
+}
+// ====
+// SMTEngine: all
```
Comment (hrkrshnn) [inline test/libsolidity/smtCheckerTests/userTypes/user_abi_2.sol:12]: The test updates are missing.

File: libsolidity/formal/SymbolicTypes.cpp [hunk 2]
```diff
@@ -233,14 +233,18 @@ pair<bool, shared_ptr<SymbolicVariable>> newSymbolicVariable(
 	bool abstract = false;
 	shared_ptr<SymbolicVariable> var;
 	frontend::Type const* type = &_type;
-	if (!isSupportedTypeDeclaration(_type))
+
+	if (auto userType = dynamic_cast<UserDefinedValueType const*>(type))
+		type = &userType->underlyingType();
```
Comment (hrkrshnn) [inline libsolidity/formal/SymbolicTypes.cpp]: Maybe it's easier to implement this by returning `newSymbolicVariable(&userType->underlyingType(), _uniqueName, _context)`?

Comment (leonardoalt) [reply libsolidity/formal/SymbolicTypes.cpp]: I'm not sure that'd be easier. `newSymbolicVariable` is used in 4 places, so we'd need this logic in all those places. I'd prefer to centralize that here.

**Comment by leonardoalt** [comment]
I *think* I managed to centralize the underlying unpacking in the fewest amount of places needed.

File: libsolidity/formal/Predicate.cpp [hunk 3]
```diff
@@ -237,7 +237,7 @@ string Predicate::formatSummaryCall(
 	auto last = first + static_cast<int>(fun->parameters().size());
 	solAssert(first >= _args.begin() && first <= _args.end(), "");
 	solAssert(last >= _args.begin() && last <= _args.end(), "");
-	auto inTypes = FunctionType(*fun).parameterTypes();
+	auto inTypes = FunctionType(*fun).parameterTypesReplacingUserTypes();
```
Comment (hrkrshnn) [inline libsolidity/formal/Predicate.cpp]: I feel that this function `parameterTypesReplacingUserTypes` does not belong in `Types.h`, but as a helper function somewhere here. What do you think?

File: libsolidity/formal/SMTEncoder.cpp [hunk 4]
```diff
@@ -406,6 +406,9 @@ void SMTEncoder::endVisit(TupleExpression const& _tuple)
 	if (_tuple.annotation().type->category() == Type::Category::Function)
 		return;
 
+	if (_tuple.annotation().type->category() == Type::Category::TypeType)
```
Comment (hrkrshnn) [inline libsolidity/formal/SMTEncoder.cpp:409]: Why is this needed?

File: libsolidity/formal/SymbolicTypes.cpp (hunk 2 above)
Comment (hrkrshnn) [inline libsolidity/formal/SymbolicTypes.cpp]:
```suggestion
	if (auto userType = dynamic_cast<UserDefinedValueType const*>(type))
		return newSymbolicVariable(&userType->underlyingType(), _uniqueName, _context);
```
This is what I meant. The rest of the code can remain unchanged.

File: libsolidity/formal/Predicate.cpp (hunk 3 above)
Comment (hrkrshnn) [reply libsolidity/formal/Predicate.cpp]: This would just be an applyMap or range-v3 transform using the function `underlyingType`.

File: libsolidity/formal/SymbolicTypes.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/SymbolicTypes.cpp]: Ah I see

File: libsolidity/formal/SMTEncoder.cpp (hunk 4 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:409]: To avoid spurious unsupported warnings

File: libsolidity/formal/Predicate.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp]: Ok!

File: libsolidity/formal/SymbolicTypes.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/SymbolicTypes.cpp]: I left the other changes in this file just cus it looked better

**Review by hrkrshnn — APPROVED** [review]

**Comment by chriseth** [comment]
Does this still need more tests (as written in the description) or is this fine to merge?

**Comment by hrkrshnn** [comment]
I think this is fine to merge. The description is old. Will merge now.
